### PR TITLE
refactor(core): convert security group filter service to TS

### PR DESCRIPTION
**Pull request by anotherchrisberry**
This started as a quick fix to stop showing the "No security group matches found" message that flashes on the initial page load when deep linked into the security group page, but then I converted something to Typescript.

### CODE REVIEW

File: app/scripts/modules/core/src/domain/ISecurityGroup.ts [hunk 1]
```diff
@@ -29,8 +31,20 @@ export interface ISecurityGroup {
...
   region?: string;
+  searchField?: string;
   stack?: string;
   type?: string;
   usages?: IUsages;
   vpcId?: string;
+  vpcName?: string;
+}
+
+export interface ISecurityGroupGroup {
```
Comment (jrsquared) [inline app/scripts/modules/core/src/domain/ISecurityGroup.ts]: i love group groups

Comment (anotherchrisberry) [reply app/scripts/modules/core/src/domain/ISecurityGroup.ts]: It's really fun to type

File: app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+          const heading = securityGroup.vpcName ?
+            `${securityGroup.region} (${securityGroup.vpcName})` :
+            securityGroup.region;
+          subSubGroups.push({
+            heading: heading,
+            vpcName: securityGroup.vpcName,
+            securityGroup: securityGroup,
+          });
+        });
+        subGroups.push( {
```
Comment (jrsquared) [inline app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts]: nice whitespace

File: app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+          heading: subKey,
+          subgroups: sortBy(subSubGroups, ['heading', 'vpcName']),
+        });
+      });
+
+      groups.push( { heading: key, subgroups: sortBy(subGroups, 'heading') } );
```
Comment (jrsquared) [inline app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts]: so much room to breathe

File: app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
+import { module } from 'angular';
+import { chain, find, forOwn, groupBy, map, sortBy } from 'lodash';
+import { Subject } from 'rxjs';
+import { BindAll } from 'lodash-decorators';
+
+import { Application } from 'core/application/application.model';
+import { ISecurityGroup, ISecurityGroupGroup } from 'core/domain';
+import { SECURITY_GROUP_FILTER_MODEL, SecurityGroupFilterModel } from './securityGroupFilter.model';
+
+@BindAll()
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts]: is this necessary?

File: app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+export class SecurityGroupFilterService {
+  public groupsUpdatedStream: Subject<ISecurityGroupGroup[]> = new Subject<ISecurityGroupGroup[]>();
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts]: either 

```
 public groupsUpdatedStream: Subject<ISecurityGroupGroup[]> = new Subject();
```

or (better IMHO)


```
 public groupsUpdatedStream = new Subject<ISecurityGroupGroup[]>();
```

File: app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    this.lastApplication = application;
+    this.groupsUpdatedStream.next(groups);
+  }
+
+}
+
+export const SECURITY_GROUP_FILTER_SERVICE = 'spinnaker.core.securityGroup.filter.service';
+module(SECURITY_GROUP_FILTER_SERVICE, [
+  SECURITY_GROUP_FILTER_MODEL,
+  require('core/filterModel/filter.model.service.js').name
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts]: import the stringy thinggy

File: app/scripts/modules/core/src/securityGroup/AllSecurityGroupsCtrl.js [hunk 7]
```diff
@@ -53,15 +50,14 @@ module.exports = angular.module('spinnaker.core.securityGroup.all.controller', [
...
-        // Timeout because the updateSecurityGroups method is debounced by 25ms
-        $timeout(() => { this.initialized = true; }, 50);
       });
     };
 
     let groupsUpdated = () => {
       $scope.$applyAsync(() => {
         $scope.groups = securityGroupFilterModel.groups;
         $scope.tags = securityGroupFilterModel.tags;
+        this.initialized = this.initialized || app.securityGroups.loaded;
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/securityGroup/AllSecurityGroupsCtrl.js]: this was one of the main problems: we were always setting the state to initialized, even if the application's security groups were not yet loaded.

File: app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts (hunk 2 above)
Comment (anotherchrisberry) [reply app/scripts/modules/core/src/securityGroup/filter/securityGroupFilter.service.ts]: you never know

File: app/scripts/modules/core/src/securityGroup/AllSecurityGroupsCtrl.js (hunk 7 above)
Comment (christopherthielen) [reply app/scripts/modules/core/src/securityGroup/AllSecurityGroupsCtrl.js]: good jorb

**Pull request merged**